src/Reduction.cpp

```diff
@@ -1,12 +1,146 @@
+#include "Var.h"
 #include "IR.h"
+#include "IREquality.h"
+#include "IROperator.h"
+#include "IRVisitor.h"
 #include "Reduction.h"
+#include "Simplify.h"
 
 namespace Halide {
 namespace Internal {
 
+namespace {
+
+/* Split AND predicate into vector of ANDs. */
+class SplitAndPredicate : public IRVisitor {
+public:
+    std::vector<Expr> predicates;
+
+    using IRVisitor::visit;
+
+    void visit(const And *op) {
+        std::vector<Expr> old;
+        old.swap(predicates);
+
+        const And *and_a = op->a.as<And>();
+        const And *and_b = op->b.as<And>();
+
+        std::vector<Expr> predicates_a;
+        if (and_a) {
+            op->a.accept(this);
+            predicates_a.swap(predicates);
+        } else {
+            predicates_a.push_back(op->a);
+        }
+
+        if (and_b) {
+            op->b.accept(this);
+        } else {
+            predicates.push_back(op->b);
+        }
+
+        predicates.insert(predicates.end(), predicates_a.begin(), predicates_a.end());
+        predicates.insert(predicates.end(), old.begin(), old.end());
+    }
+};
+
+std::vector<Expr> split_predicate_helper(Expr pred) {
```

Comment:
Suggested simpler implementation -- In the past I've done this sort of thing like so:

    void split_predicate_helper(Expr pred, std::vector<Expr> &result) {
       if (const And *a = pred.as<And>()) {
           split_predicate_helper(a->a, result);
           split_predicate_helper(a->b, result);
       } else {
           result.push_back(pred);
       }
    }

which you then use like so:

    std::vector<Expr> terms;
    split_predicate_helper(pred, terms);